### PR TITLE
Add libsodium resolver (up-to-date)

### PR DESCRIPTION
**Pull request by expenses**
This is an updated fork of https://github.com/mcginty/snow/pull/20. I haven't yet applied any suggestions from code review.

### CODE REVIEW

**Comment by mcginty** [comment]
Oh yay, nice!

For summary's sake, I think the remaining work is:

* Implement the `Random` trait for sodiumoxide's `randombytes()`
* Work out our way of guaranteeing libsodium's `init()` has been called, either with the help of something like lazy_static or once_cell, or something as simple as calling the initializer every time a `Session` is constructed. Quickly glancing at libsodium's documentation, it seems smart enough to not re-initialize every time the initialize method is called.

**Comment by expenses** [comment]
Ok, I've added the rng, but I'm not sure where the best place to put the init function. According to https://download.libsodium.org/doc/usage/ it's perfectly safe to call multiple times.

**Comment by mcginty** [comment]
@expenses While it's clearly redundant, since the init function doesn't seem particularly costly and is safe to call many times, it seems safest to put a call to int inside the constructor of each of the algorithm structures (eg. inside of `SodiumDh25519::default()`).

**Comment by expenses** [comment]
@mcginty done.

File: Cargo.toml [hunk 1]
```diff
@@ -48,6 +50,9 @@ pqcrypto-traits = { version = "0.3", optional = true }
 
 # ring crypto proivder
 ring = { version = "^0.16.2", optional = true, features = ["std"] }
+# libsodium crypto provider
+sodiumoxide = { version = "0.2.5", optional = true }
+byteorder = { version = "1.3.2", optional = true }
```
Comment (mcginty) [inline Cargo.toml]: Unless necessary for compatibility or security, can we set these to just "0.2" and "1.3" to give users of the library a bit more flexibility to reduce build times and dependencies?

File: src/resolvers/libsodium.rs [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    }
+
+    fn dh(&self, pubkey: &[u8], out: &mut [u8]) -> Result<(), ()> {
+        let pubkey = sodium_curve25519::GroupElement::from_slice(&pubkey[0..32])
+            .expect("Can't construct public key for Dh25519");
+        let result = sodium_curve25519::scalarmult(&self.privkey, &pubkey);
+
+        match result {
+            Ok(ref buf) => {
+                out[..32].copy_from_slice(&buf[0..32]);
```
Comment (mcginty) [inline src/resolvers/libsodium.rs]: I know it's non-canonical Rust, but let's use the `copy_slices!` macro for consistency until we change it library-wide in one fell swoop:

```rust
copy_slices!(&buf, out);
```

There are a couple more instances of this below.

**Comment by expenses** [comment]
I've made those changes. I'd rather not rebase and squash the commits, as it's easier for you to use the 'Squash and merge' button.

**Comment by mcginty** [comment]
After merging, I realized that this PR didn't implement the `libsodium-accelerated` feature, it just declared it. After enabling it and actually running the tests against the libsodium resolver, some tests fail (specifically, tests using a `[0u8; 32]` public key) because libsodium has made the controversial choice of validating keys [which is not what Noise should do](https://moderncrypto.org/mail-archive/noise/2017/000971.html).

Unless there are any other ideas like simply providing a caveat in documentatoin for the libsodium resolver, I'm inclined to add a workaround to the code to allow for all-zero public keys to maintain correctness with Noise.

**Comment by tarcieri** [comment]
The simplest workaround I can think of short of actually aborting the handshake if a point fails to decode is substituting some well-known point (e.g. the generator point)

**Comment by mcginty** [comment]
@tarcieri sorry, I'm not sure I understand - do you mean substituting a [0u8; 32] public key with the generator point? Wouldn't that produce outputs incompatible with other Noise implementations?

**Comment by tarcieri** [comment]
> sorry, I'm not sure I understand - do you mean substituting a [0u8; 32] public key

You need to handle not just a point consisting of all zeroes, but all of the low-order points listed in the "How do I validate Curve25519 public keys?" here:

https://cr.yp.to/ecdh.html#validate

All of these have a scalarmult/D-H output of 0 regardless of the value of the scalar.

I don't think you need to hardcode this list though: libsodium already did, so what you need to handle is a point validation failure coming from libsodium.

> ...with the generator point? Wouldn't that produce outputs incompatible with other Noise implementations?

Ok, forget I said that, it was a silly suggestion 😉

Instead when libsodium fails to decode a point, you can skip scalar multiplication entirely and use `[0u8; 32]` as the scalarmult/D-H output (since that's what it would be anyway).

That should be fully compatible with other Noise implementations.

**Comment by mcginty** [comment]
Thanks for the response @tarcieri! Are there any timing attack issues with that approach? I'm going to make an open issue for this, add a caveat to the README, and in the meantime not use all-zero points in the general tests.

**Comment by tarcieri** [comment]
There aren’t any interactions with the private scalar in that case, so there aren’t any concerns about timing variability.

Note that the May the Fourth paper specifically suggests the approach of eagerly rejecting these points specifically to avoid potential microarchitectural attacks involving interactions between the bad points and the private scalar (see Section 5):

https://eprint.iacr.org/2017/806.pdf

**Pull request merged**